File: java/leetcode_grind/Day333.java
package leetcode_grind;

import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.TreeMap;
import java.util.function.Function;

public class Day333 {
    // https://leetcode.com/problems/number-of-flowers-in-full-bloom/description
    static class Solution {
        public int[] fullBloomFlowersPriorityQueue(int[][] flowers, int[] people) {
            var sortedPeople = people.clone();
            Arrays.sort(sortedPeople);
            Arrays.sort(flowers, (f1, f2) -> f1[0] - f2[0]);

            var personToBloomedFlowers = new HashMap<Integer, Integer>();
            var pq = new PriorityQueue<Integer>();
            var i = 0;
            for (var person : sortedPeople) {
                while (i < flowers.length && flowers[i][0] <= person) {
                    pq.add(flowers[i][1]);
                    i++;
                }
                while (!pq.isEmpty() && pq.peek() < person) {
                    pq.poll();
                }
                personToBloomedFlowers.put(person, pq.size());
            }

            i = 0;
            var ans = new int[people.length];
            for (var person : people) {
                ans[i++] = personToBloomedFlowers.get(person);
            }
            return ans;
        }

        public int[] fullBloomFlowersLineSweep(int[][] flowers, int[] people) {
            var sortedMap = new TreeMap<Integer, Integer>();
            sortedMap.put(0, 0);
            for (var flower : flowers) {
                var start = flower[0];
                var end = flower[1] + 1;
                sortedMap.put(start, sortedMap.getOrDefault(start, 0) + 1);
                sortedMap.put(end, sortedMap.getOrDefault(end, 0) - 1);
            }

            var i = 0;
            var curr = 0;
            var positions = new int[sortedMap.size()];
            var prefix = new int[sortedMap.size()];
            for (var entry : sortedMap.entrySet()) {
                curr += entry.getValue();
                positions[i] = entry.getKey();
                prefix[i] = curr;
                i++;
            }

            Function<Integer, Integer> upperBound = (target) -> {
                var lo = 0;
                var hi = positions.length;
                while (lo < hi) {
                    var mid = (lo + hi) / 2;
                    if (positions[mid] <= target) {
                        lo = mid + 1;
                    } else {
                        hi = mid;
                    }
                }
                return lo;
            };

            var ans = new int[people.length];
            for (int j = 0; j < people.length; j++) {
                var idx = upperBound.apply(people[j]) - 1;
                ans[j] = prefix[idx];
            }

            return ans;
        }

        public int[] fullBloomFlowersBinarySearch(int[][] flowers, int[] people) {
            var startBloom = new int[flowers.length];
            var endBloom = new int[flowers.length];
            var i = 0;
            for (var flower : flowers) {
                startBloom[i] = flower[0];
                endBloom[i] = flower[1] + 1;
                i++;
            }
            Arrays.sort(startBloom);
            Arrays.sort(endBloom);

            var binSearch = new Object() {
                int apply(int[] arr, int target) {
                    var lo = 0;
                    var hi = arr.length;
                    while (lo < hi) {
                        var mid = (lo + hi) / 2;
                        if (arr[mid] <= target) {
                            lo = mid + 1;
                        } else {
                            hi = mid;
                        }
                    }
                    return lo;
                }
            };

            var ans = new int[people.length];
            for (int j = 0; j < people.length; j++) {
                var person = people[j];
                var start = binSearch.apply(startBloom, person);
                var end = binSearch.apply(endBloom, person);
                ans[j] = start - end;
            }
            return ans;
        }
    }
}

File: src/leetcode_grind/day_333.rs
// https://leetcode.com/problems/number-of-flowers-in-full-bloom/description/
pub fn full_bloom_flowers_heap(flowers: Vec<Vec<i32>>, people: Vec<i32>) -> Vec<i32> {
    use std::cmp::Reverse;
    use std::collections::BinaryHeap;
    use std::collections::HashMap;

    let mut flowers = flowers;
    flowers.sort();

    let mut sorted_people = people.clone();
    sorted_people.sort();

    let mut dic = HashMap::new();
    let mut heap = BinaryHeap::new();

    let mut i = 0;
    for person in sorted_people {
        while i < flowers.len() && flowers[i][0] <= person {
            heap.push(Reverse(flowers[i][1]));
            i += 1;
        }

        while !heap.is_empty() && heap.peek().unwrap().0 < person {
            heap.pop();
        }

        dic.insert(person, heap.len() as i32);
    }

    let mut ans = vec![0; people.len()];
    for j in 0..people.len() {
        ans[j] = dic[&people[j]];
    }

    ans
}

pub fn full_bloom_flowers_binary_search(flowers: Vec<Vec<i32>>, people: Vec<i32>) -> Vec<i32> {
    fn binary_search(positions: &Vec<i32>, target: i32) -> usize {
        let mut left = 0;
        let mut right = positions.len() as i32;

        while left < right {
            let mid = (left + right) / 2;
            if positions[mid as usize] <= target {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        left as usize
    }

    use std::collections::BTreeMap;
    let mut difference = BTreeMap::new();
    difference.insert(0, 0);

    for flower in flowers {
        let start = flower[0];
        let end = flower[1] + 1;

        *difference.entry(start).or_insert(0) += 1;
        *difference.entry(end).or_insert(0) -= 1;
    }

    let mut positions = vec![];
    let mut prefix = vec![];
    let mut curr = 0;

    for &key in difference.keys() {
        positions.push(key);
        curr += difference[&key];
        prefix.push(curr);
    }

    let mut ans = vec![0; people.len()];
    for j in 0..people.len() {
        let i = binary_search(&positions, people[j]) - 1;
        ans[j] = prefix[i];
    }
    ans
}

pub fn full_bloom_flowers_simpler_binary_search(
    flowers: Vec<Vec<i32>>,
    people: Vec<i32>,
) -> Vec<i32> {
    fn binary_search(arr: &Vec<i32>, target: i32) -> i32 {
        let mut lo = 0;
        let mut hi = arr.len() as i32;
        while lo < hi {
            let mid = (lo + hi) / 2;
            if arr[mid as usize] <= target {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        lo
    }

    let mut starts = vec![];
    let mut ends = vec![];

    for flower in flowers {
        let start = flower[0];
        let end = flower[1];
        starts.push(start);
        ends.push(end + 1);
    }
    starts.sort();
    ends.sort();

    let mut ans = vec![];
    for person in people {
        let s = binary_search(&starts, person);
        let e = binary_search(&ends, person);
        ans.push(s - e);
    }
    ans
}
